File: cpp/cpp_100-999/648.cc
// C++
// leetcode 648
// https://leetcode.cn/problems/replace-words/
// 字典树

class Solution {
public:
	// int N = 100009;
	int trie[100009][26];
	int count[100009];
	int idx = 0;

	void insert(string s) {
		int p = 0;
		for(int i = 0; i < s.size(); i++) {
			int u = s[i] - 'a';
			if(trie[p][u] == 0)
				trie[p][u] = ++idx;
			p = trie[p][u];
		}
		count[p]++;
        // cout<<s<<" inserted"<<endl;
	}

	string searchPre(string s) {
		int p = 0;
		for(int i = 0; i < s.size(); i++) {
			int u = s[i] - 'a';
			if(trie[p][u] == 0) {
				if(i == 0 || count[p] == 0) {
					return s;
				}
			}

			if(count[p] != 0) {
				return s.substr(0, i);
			}
			p = trie[p][u];
		}
		return s;
	}

    string replaceWords(vector<string>& dictionary, string sentence) {
    	memset(trie, 0, sizeof(trie));
    	for(auto &s : dictionary) {
    		insert(s);
    	}
    	int n = sentence.size();
    	string res = "";
    	for(int i = 0; i < n; ) {
    		int j = i;
    		while(j < n && sentence[j] != ' ') {
    			j++;
    		}
    		// string cur = sentence.substr(i, j - i);
    		// cout<<"searching "<<cur<<endl;
    		res += searchPre(sentence.substr(i, j - i));
    		if(j != n) {
    			res += " ";
    		}
    		i = j + 1;
    	}
    	return res;
    }
};

File: Go/Go_100-999/648.go
// Go
// leetcode 648
// https://leetcode.cn/problems/replace-words/

type TrieNode struct {
	trie  [100009][26]int
	count [100009]int
	idx   int
}

func (t *TrieNode) insert(s string) {
	p := 0
	for _, c := range s {
		u := int(c - 'a')
		if t.trie[p][u] == 0 {
			t.idx++
			t.trie[p][u] = t.idx
		}
		p = t.trie[p][u]
		if t.count[p] > 0 {
			return
		}
	}
	t.count[p]++
}

func (t *TrieNode) searchPre(s string) string {
	p := 0
	// fmt.Println("in: ", s)
	for i, c := range s {
		u := int(c - 'a')
		if t.trie[p][u] == 0 && (i == 0 || t.count[p] == 0) {
			// fmt.Println("out: ", s)
			return s
		}
		if t.count[p] != 0 {
			// fmt.Println("out: ", s[0 : i])
			return s[0:i]
		}

		p = t.trie[p][u]
	}
	// fmt.Println("out: ", s)
	return s
}

func (t *TrieNode) clear() {
	i := t.idx
	for i >= 0 {
		t.count[i] = 0
		for j := 0; j < 26; j++ {
			t.trie[i][j] = 0
		}
		i--
	}
	t.idx = 0
}

var tn *TrieNode

func replaceWords(dictionary []string, sentence string) string {
	// 初始化字典树
	if tn == nil {
		tn = &TrieNode{}
	} else {
		tn.clear()
	}

	// 将前缀录入
	for _, s := range dictionary {
		tn.insert(s)
	}

	n := len(sentence)
	res := ""
	for i := 0; i < n; {
		j := i
		for j < n && sentence[j] != ' ' {
			j++
		}
		res += tn.searchPre(sentence[i:j])
		if j != n {
			res += " "
		}
		i = j + 1
	}
	return res
}
